src/miral/keymap.cpp

```diff
@@ -244,3 +245,166 @@ void miral::Keymap::set_keymap(std::string const& keymap)
+                                                        nullptr,
+                                                        G_DBUS_CALL_FLAGS_NONE,
+                                                        G_MAXINT,
+                                                        nullptr,
+                                                        &error)})
+    {
+        Variant const unwrap{g_variant_get_child_value(result, 0)};
+        Variant const unwrap2{g_variant_get_child_value(unwrap, 0)};
+
+        if (g_variant_is_of_type(unwrap2, G_VARIANT_TYPE_STRING))
+        {
+            gsize len = 0;
+            return g_variant_get_string(unwrap2, &len);
+        }
+    }
+
+    if (error)
+    {
+        mir::log_info("Dbus error=%s, dest=%s, object_path=%s, properties_interface=%s, method_name=%s, interface_name=%s",
+                      error->message, bus_name, object_path, properties_interface, method_name, interface_name);
+        g_error_free(error);
+    }
+
+    return std::nullopt;
+}
+
+auto read_keymap(Connection const& connection) -> std::optional<std::string>
+{
+    if (auto const layout = read_entry(connection, "X11Layout"))
+    {
+        auto const variant = read_entry(connection, "X11Variant");
+        auto const options = read_entry(connection, "X11Options");
+
+        return *layout + '+' + variant.value_or("") + '+' + options.value_or("");
+    }
+
+    return std::nullopt;
+}
+
+template<typename T>
+void callback_thunk(GDBusConnection*, const gchar*, const gchar*, const gchar*, const gchar*, GVariant *parameters, gpointer self)
+{
+    auto const self_ptr = static_cast<T*>(self);
+    self_ptr->callback(parameters);
+}
+}
+
+auto miral::Keymap::system_locale1() -> Keymap
+{
+    struct SystemLocalSelf : Self
+    {
+        explicit SystemLocalSelf(Connection const&& connection) :
+            Self{read_keymap(connection).value_or("us")},
+            watch_id{g_dbus_connection_signal_subscribe(
+                connection,
+                sender,
+                properties_interface,
+                signal_name,
+                object_path,
+                nullptr,
+                G_DBUS_SIGNAL_FLAGS_NONE,
+                callback_thunk<SystemLocalSelf>,
+                this,
+                nullptr)},
+            connection{std::move(connection)},
+            main_loop{MainLoop::the_main_loop()}
+        {
+        }
+
+        ~SystemLocalSelf() override
+        {
+            g_dbus_connection_signal_unsubscribe(connection, watch_id);
+        }
+
+        guint watch_id = 0;
+        Connection const connection;
+        std::shared_ptr<MainLoop> const main_loop;
+
+        void callback(GVariant* parameters)
+        {
+            std::optional<std::string> layout;
+            std::optional<std::string> options;
+            std::optional<std::string> variant;
+
+            const char* _;
+            GVariant* changed_properties;
```

Comment:
Can use `g_autoptr(GVariant) changed_properties = nullptr;` to avoid the use of the g_variant_free later (and same for `invalidated_properties`).